src/hotspot/share/opto/predicates.hpp

```diff
@@ -283,14 +283,46 @@ class TemplateAssertionPredicateExpression : public StackObj {
   Opaque4Node* clone(const TransformStrategyForOpaqueLoopNodes& transform_strategy, Node* new_ctrl, PhaseIdealLoop* phase);
 
  public:
-  // Is 'n' a node that could be part of a Template Assertion Predicate Expression (i.e. could be found on the input
-  // chain of a Template Assertion Predicate Opaque4Node up to and including the OpaqueLoop* nodes)?
-  static bool maybe_contains(const Node* n) {
-    const int opcode = n->Opcode();
-    return (opcode == Op_OpaqueLoopInit ||
-            opcode == Op_OpaqueLoopStride ||
-            n->is_Bool() ||
-            n->is_Cmp() ||
+  Opaque4Node* clone(Node* new_ctrl, PhaseIdealLoop* phase);
+  Opaque4Node* clone_and_replace_init(Node* new_init, Node* new_ctrl,PhaseIdealLoop* phase);
+  Opaque4Node* clone_and_replace_init_and_stride(Node* new_init, Node* new_stride, Node* new_ctrl, PhaseIdealLoop* phase);
+};
+
+// Class to represent a node being part of a Template Assertion Predicate Expression.
+//
+// The expression itself can belong to no, one, or two Template Assertion Predicates:
+// - None: This node is already dead (i.e. we replaced the Bool condition of the Template Assertion Predicate).
+// - Two: A OpaqueLoopInitNode could be part of two Template Assertion Predicates.
+// - One: In all other cases.
+class TemplateAssertionPredicateExpressionNode : public StackObj {
+  Node* const _node;
+
+ public:
+  explicit TemplateAssertionPredicateExpressionNode(Node* node) : _node(node) {
+    assert(is_valid(node), "must be valid");
+  }
+  NONCOPYABLE(TemplateAssertionPredicateExpressionNode);
+
+ private:
+  static bool find_opaque_loop_nodes(Node* node);
+  static void push_non_null_inputs(Unique_Node_List& list, const Node* node);
+  static bool is_template_assertion_predicate(Node* node);
+  static void push_outputs(Unique_Node_List& list, const Node* node);
+
+ public:
+  // Check whether the provided node is part of a Template Assertion Predicate Expression or not.
+  static bool is_valid(Node* node) {
+    return find_opaque_loop_nodes(node);
+  }
+
+  // Check if the opcode of node could be found in a Template Assertion Predicate Expression.
+  // This also provides a fast check whether a node is unrelated.
+  static bool valid_opcode(const Node* node) {
```

Comment:
It's always a good sign when you point out the names I'm also not quite satisfied to start with - so that's an indicator to change it :-) 

I tried to be concise here but then there are really no good and precise options. Let's go with `is_maybe_in_expression()` and `is_in_expression()` as also suggested above.